Dynamo+LTC: handle inplace ops

shunting314 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #75359

For some models in torchbench (e.g. pyhpc_isoneutral_mixing), dynamo will generate Fx graphs that has side effects . Those graphs may
- return an empty tuple
- change tensors passed in as forward method arguments in-place

This makes the Dynamo+LTC integration fail since we extract compiled graph based on the lazy tensors returned from the forward method. From an empty tuple, we extract nothing.

To solve this problem, we extract compile graph from `union(argument lazy tensors, returned lazy tensors)` instead. The inplace mutations applied to argument lazy tensors will be captured this way.


Test plan:
```
pytest test/lazy/test_extract_compiled_graph.py
```

```
LTC_TS_CUDA=1 gpui time python torchbench.py --speedup-ltc -dcuda --nvfuser --randomize-input --only pyhpc_isoneutral_mixing
```

Differential Revision: [D35478799](https://our.internmc.facebook.com/intern/diff/D35478799)

shunting314 committed:
Dynamo+LTC: handle inplace ops

[ghstack-poisoned]

wconstab commented on torch/_lazy/extract_compiled_graph.py:
```diff
@@ -178,12 +179,21 @@ def extract_compiled_graph(model: fx.GraphModule, example_inputs) -> Callable:
+        assert len(res) == len(args_and_out)
+        for i, arg in enumerate(args):
+            # only copy those tensors that get inplace updated
+            if arg is not res[i]:
```
what does this 'is' check actually check?

shunting314 replied:
> what does this 'is' check actually check?

So we compiled a graph for `union(argument lazy tensors, returned lazy tensors)`. That graph will return (after duplication in case some tensors are duplicated) all the argument tensors at the beginning of the list. There are 2 cases
- some of those argument tensors may get inplace updated. In that case, the graph will return a new copy of the tensor. We need copy the content back to the argument
- other arguments may not get any inplace update. In that case, we can skip the copying.

Removing the 'is' check should also work but make things a bit slower.

shunting314 committed:
Update on "Dynamo+LTC: handle inplace ops"


For some models in torchbench (e.g. pyhpc_isoneutral_mixing), dynamo will generate Fx graphs that has side effects . Those graphs may
- return an empty tuple
- change tensors passed in as forward method arguments in-place

This makes the Dynamo+LTC integration fail since we extract compiled graph based on the lazy tensors returned from the forward method. From an empty tuple, we extract nothing.

To solve this problem, we extract compile graph from `union(argument lazy tensors, returned lazy tensors)` instead. The inplace mutations applied to argument lazy tensors will be captured this way.


Test plan:
```
pytest test/lazy/test_extract_compiled_graph.py
```

```
LTC_TS_CUDA=1 gpui time python torchbench.py --speedup-ltc -dcuda --nvfuser --randomize-input --only pyhpc_isoneutral_mixing
```

[ghstack-poisoned]

shunting314 commented:
@shunting314 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35478799).
